File: packages/adb/src/test/lib/merkle-tree/balance.test.ts
import { rotateRight, rotateLeft, rebalanceTree } from '../../../lib/merkle-tree';
import { node, leaf, expectNode } from './merkle-verify';

describe('balance', () => {

    test('right rotation', () => {
        /*
         * Before rotation:
         *     A
         *    / \
         *   B   D
         *  / \
         * C   E
         * 
         * After rotation:
         *     B
         *    / \
         *   C   A
         *      / \
         *     E   D
         */
        const C = leaf('C', 100);
        const E = leaf('E', 200);
        const D = leaf('D', 300);
        const B = node(C, E);
        const A = node(B, D);

        const result = rotateRight(A);
        expectNode(expect.getState().currentTestName!, result, {
            left: C,
            right: {
                left: E,
                right: D,                    
            },
        });
    });

    test('left rotation', () => {
        /*
         * Before rotation:
         *     A
         *    / \
         *   B   C
         *      / \
         *     D   E
         * 
         * After rotation:
         *     C
         *    / \
         *   A   E
         *  / \
         * B   D
         */
        const B = leaf('B', 100);
        const D = leaf('D', 200);
        const E = leaf('E', 300);
        const C = node(D, E);
        const A = node(B, C);

        const result = rotateLeft(A);
        expectNode(expect.getState().currentTestName!, result, {
            left: {
                left: B,
                right: D,
            },
            right: E,
        });
    });

    test('simple balanced tree requires no rotation', () => {
        /*
         * Balanced tree:
         *     A
         *    / \
         *   B   C
         */
        const C = leaf('C');
        const B = leaf('B');
        const A = node(B, C);

        const result = rebalanceTree(A);
        expectNode(expect.getState().currentTestName!, result, {
            left: B,
            right: C,
        });
    });

    test('slightly left heavy tree requires no rotation', () => {
        /*
         * Left heavy
         *     A
         *    / \
         *   B   E
         *  / \ 
         * C  D
         */
        const E = leaf('E');
        const D = leaf('D');
        const C = leaf('C');
        const B = node(C, D);
        const A = node(B, E);

        const result = rebalanceTree(A);
        expectNode(expect.getState().currentTestName!, result, {
            left: B,
            right: E,
        });
    });

    test('slightly right heavy tree requires rotation', () => {
        /*
         * Right heavy
         *     A
         *    / \
         *   B   C
         *      / \
         *     D   E
         */
        const B = leaf('B');
        const D = leaf('D');
        const E = leaf('E');
        const C = node(D, E);
        const A = node(B, C);

        const result = rebalanceTree(A);
        expectNode(expect.getState().currentTestName!, result, {
            left: {
                left: B,
                right: D,
            },
            right: E,
        });
    });

    test(`left heavy tree requires right rotation`, () => {
        /*
         * Heavily left-heavy tree:
         *         A
         *        / \
         *       B   C
         *      / \
         *     D   E
         *    / \
         *   F   G
         * 
         * After right rotation:
         *       B
         *      / \
         *     D   A
         *    / \ / \
         *   F G E   C
         */
        const F = leaf('F');
        const G = leaf('G');
        const D = node(F, G);
        const E = leaf('E');
        const B = node(D, E);
        const C = leaf('C');
        const A = node(B, C);

        const result = rebalanceTree(A);
        
        // Verify the structure after rebalancing
        expectNode(expect.getState().currentTestName!, result, {
            left: D,
            right: {
                left: E,
                right: C,
            },
        });
    });

    test(`right heavy tree requires left rotation`, () => {
        /*
         * Heavily right-heavy tree:
         *         A
         *        / \
         *       B   C
         *          / \
         *         D   E
         *            / \
         *           F   G
         * 
         * After left rotation:
         *       C
         *      / \
         *     A   E
         *    / \ / \
         *   B  D F  G
         */
        const B = leaf('B');
        const D = leaf('D');
        const F = leaf('F');
        const G = leaf('G');
        const E = node(F, G);
        const C = node(D, E);
        const A = node(B, C);

        const result = rebalanceTree(A);
        
        // Verify the structure after rebalancing
        expectNode(expect.getState().currentTestName!, result, {
            left: {
                left: B,
                right: D,
            },
            right: E,
        });
    });
});

File: packages-zig/merkle-tree-zig/src/test/balance.test.zig
//
// Tests for rotateRight, rotateLeft and rebalanceTree (port of src/test/balance.test.ts).
//

const std = @import("std");
const merkle_tree_zig = @import("merkle-tree-zig");
const merkle_verify = @import("merkle-verify.zig");
const merkle_tree = merkle_tree_zig.merkle_tree;
const leaf = merkle_verify.leaf;
const node = merkle_verify.node;
const expectNode = merkle_verify.expectNode;

test "right rotation" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    //
    // Before rotation:
    //     A
    //    / \
    //   B   D
    //  / \
    // C   E
    //
    // After rotation:
    //     B
    //    / \
    //   C   A
    //      / \
    //     E   D
    //
    const nodeC = try leaf(allocator, "C", 100);
    const nodeE = try leaf(allocator, "E", 200);
    const nodeD = try leaf(allocator, "D", 300);
    const nodeB = try node(allocator, nodeC, nodeE);
    const nodeA = try node(allocator, nodeB, nodeD);

    const result = try merkle_tree.rotateRight(allocator, nodeA);
    try expectNode(allocator, result, "(C,(E,D))");
}

test "left rotation" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    //
    // Before rotation:
    //     A
    //    / \
    //   B   C
    //      / \
    //     D   E
    //
    // After rotation:
    //     C
    //    / \
    //   A   E
    //  / \
    // B   D
    //
    const nodeB = try leaf(allocator, "B", 100);
    const nodeD = try leaf(allocator, "D", 200);
    const nodeE = try leaf(allocator, "E", 300);
    const nodeC = try node(allocator, nodeD, nodeE);
    const nodeA = try node(allocator, nodeB, nodeC);

    const result = try merkle_tree.rotateLeft(allocator, nodeA);
    try expectNode(allocator, result, "((B,D),E)");
}

test "simple balanced tree requires no rotation" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    const nodeC = try leaf(allocator, "C", 100);
    const nodeB = try leaf(allocator, "B", 100);
    const nodeA = try node(allocator, nodeB, nodeC);

    const result = try merkle_tree.rebalanceTree(allocator, nodeA);
    try std.testing.expectEqual(nodeA, result);
    try expectNode(allocator, result, "(B,C)");
}

test "slightly left heavy tree requires no rotation" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    const nodeE = try leaf(allocator, "E", 100);
    const nodeD = try leaf(allocator, "D", 100);
    const nodeC = try leaf(allocator, "C", 100);
    const nodeB = try node(allocator, nodeC, nodeD);
    const nodeA = try node(allocator, nodeB, nodeE);

    const result = try merkle_tree.rebalanceTree(allocator, nodeA);
    try expectNode(allocator, result, "((C,D),E)");
}

test "slightly right heavy tree requires rotation" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    const nodeB = try leaf(allocator, "B", 100);
    const nodeD = try leaf(allocator, "D", 100);
    const nodeE = try leaf(allocator, "E", 100);
    const nodeC = try node(allocator, nodeD, nodeE);
    const nodeA = try node(allocator, nodeB, nodeC);

    const result = try merkle_tree.rebalanceTree(allocator, nodeA);
    try expectNode(allocator, result, "((B,D),E)");
}

test "left heavy tree requires right rotation" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    const nodeF = try leaf(allocator, "F", 100);
    const nodeG = try leaf(allocator, "G", 100);
    const nodeD = try node(allocator, nodeF, nodeG);
    const nodeE = try leaf(allocator, "E", 100);
    const nodeB = try node(allocator, nodeD, nodeE);
    const nodeC = try leaf(allocator, "C", 100);
    const nodeA = try node(allocator, nodeB, nodeC);

    const result = try merkle_tree.rebalanceTree(allocator, nodeA);

    // Verify the structure after rebalancing
    try expectNode(allocator, result, "((F,G),(E,C))");
}

test "right heavy tree requires left rotation" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    const nodeB = try leaf(allocator, "B", 100);
    const nodeD = try leaf(allocator, "D", 100);
    const nodeF = try leaf(allocator, "F", 100);
    const nodeG = try leaf(allocator, "G", 100);
    const nodeE = try node(allocator, nodeF, nodeG);
    const nodeC = try node(allocator, nodeD, nodeE);
    const nodeA = try node(allocator, nodeB, nodeC);

    const result = try merkle_tree.rebalanceTree(allocator, nodeA);

    // Verify the structure after rebalancing
    try expectNode(allocator, result, "((B,D),(F,G))");
}

test "rebalanceTree throws for a node without children" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    const lonely = try leaf(allocator, "A", 100);
    try std.testing.expectError(error.Thrown, merkle_tree.rebalanceTree(allocator, lonely));
    try std.testing.expectError(error.Thrown, merkle_tree.rotateLeft(allocator, lonely));
    try std.testing.expectError(error.Thrown, merkle_tree.rotateRight(allocator, lonely));
}
